WIP_Add Django Debug Toolbar

FroggyFlox opened the pull request:
Initial commit adding Django Debug Toolbar #2939 

- Add django-debug-toolbar package to the optional `dev` dependency group
- Add `--dev` flag to build.sh to run `poetry install` with that flag
- Add DjDTB to settings.py
- Add DjDTB urls

This pull request is created with the only purpose to illustrate what would need to be implemented to use Django Debug Toolbar.
Ideally, all changes to `settings.py` and `urls.py` would need to be scripted as well so that no manual configuration would be required.

To run this branch, first checkout, then build from source with the new `--dev` flag. For instance, I run the following:
```
cd /opt/rockstor
systemctl stop rockstor rockstor-pre rockstor-build rockstor-bootstrap postgresql && systemctl start postgresql && ./build.sh --dev && poetry run initrock && systemctl enable --now rockstor-bootstrap && poetry run debug-mode ON
```

Note that `DEBUG` must be True for the toolbar to show up.

Hooverdan96 commented on src/rockstor/settings.py:
```diff
@@ -470,3 +472,16 @@
+]
+
+# https://django-debug-toolbar.readthedocs.io/en/stable/configuration.html
+DEBUG_TOOLBAR_CONFIG = {
+    # Update to the latest AJAX request
+    # Without this, we can only catch the initial request
+    # which is not helpful in most of our cases
+    "UPDATE_ON_FETCH": True
+}
```
```suggestion
DEBUG_TOOLBAR_CONFIG = {
    # Update to the latest AJAX request
    # Without this, we can only catch the initial request
    # which is not helpful in most of our cases
    "UPDATE_ON_FETCH": True
    # If using htmx, to retain toolbar handle
    # through page renders add this
    "ROOT_TAG_EXTRA_ATTRS": "hx-preserve"    
}
```

If the project is contemplating to use htmx (#2735 and #2821) then this would also be required (among other things of course to enable htmx overall), according to:
https://django-debug-toolbar.readthedocs.io/en/stable/tips.html#working-with-htmx-and-turbo

FroggyFlox commented:
@phillxnet , I have now tried a different approach to "trigger" the installation of the dev dependencies and the toolbar itself into Django. The idea was to try to "streamline/simplify" the process for the user/developer and cut down the number of steps. I'm anxious about affecting "normal" builds and installs, though, so I would appreciate your feedback; hopefully my commits here won't affect an RPM build, install, and update.

## Testing
Ensure all traces from previous/current install are wiped
```
systemctl stop rockstor rockstor-pre rockstor-build rockstor-bootstrap
rm -vf /usr/lib/systemd/system/rockstor*
systemctl daemon-reload
# DJANGO_DEBUG should not be set but let's make sure of it
unset $DJANGO_DEBUG

cd /opt/rockstor
rm -rf .venv/
rm -rf static/
```

Now, we can build
Manually edit `.env` to include `DJANGO_DEBUG=True`

Given this would be the very first run of the script, we can't rely on
rockstor-build.service to run build.sh with the ENV from .env so we need to set it manually here
```
systemctl stop postgresql && systemctl start postgresql && DJANGO_DEBUG=True ./build.sh && poetry run initrock && systemctl enable --now rockstor-bootstrap
```

Open browser and visit Rockstor's IP: the toolbar shows up

To test rebuild after some development edits:
```
# Wipe existing venv
cd /opt/rockstor
rm -rf .venv/
rm -rf static/
systemctl stop postgresql rockstor rockstor-pre rockstor-build rockstor-bootstrap && systemctl start rockstor-build && systemctl enable --now rockstor-bootstrap
```
Open browser and visit Rockstor's IP: the toolbar shows up

To test a normal build still works without installing the Debug Toolbar
Manually edit `.env` to comment `DJANGO_DEBUG=True`
```
cd /opt/rockstor
rm -rf .venv/
rm -rf static/
systemctl stop postgresql rockstor rockstor-pre rockstor-build rockstor-bootstrap && systemctl start rockstor-build && systemctl enable --now rockstor-bootstrap
```
Open browser and visit Rockstor's IP: the toolbar DOES NOT show up

FroggyFlox commented:
Currently exploring the option of NOT having the installation of Django Debug Toolbar conditioned on a `dev` flag of sort but instead installing it by default (setting it as a normal dependency).

```
buildvm:/opt/rockstor # poetry add django-debug-toolbar
Using version ^4.4.6 for django-debug-toolbar

Updating dependencies
Resolving dependencies... (0.2s)

Package operations: 1 install, 0 updates, 0 removals

  • Installing django-debug-toolbar (4.4.6)

Writing lock file
```

Then, clear current install and re-build:
```
buildvm:/opt/rockstor # rm -rf .initrock .venv/ static/

buildvm:/opt/rockstor # systemctl stop postgresql && systemctl start postgresql && ./build.sh && poetry run initrock && systemctl enable --now rockstor-bootstrap

# confirm DEBUG is false
buildvm:/opt/rockstor # poetry run debug-mode
currently debug flag is False
```

Go to webUI, create user account (first time login): no toolbar is showed on pages (browsed to a random handful ones).

Now, let's turn debug-mode ON:
```
buildvm:/opt/rockstor # poetry run debug-mode ON
DEBUG flag is now set to True
```

Visit webUI: the toolbar does NOT appear...
Try a machine reboot: the toolbar still does NOT appear...

Try a complete re-do:
```
buildvm:/opt/rockstor # systemctl stop postgresql && systemctl start postgresql && ./build.sh && poetry run initrock && systemctl enable --now rockstor-bootstrap
```

Visit webUI: the toolbar DOES appear... not sure why, though... still looking.

FroggyFlox commented:
With the latest commit (https://github.com/rockstor/rockstor-core/pull/2940/commits/4faa05022e225af0cf57846ec661a0a4c0b5968b), the process would thus be as follows:

```
systemctl stop rockstor rockstor-pre rockstor-build rockstor-bootstrap
rm -vf /usr/lib/systemd/system/rockstor*
systemctl daemon-reload

cd /opt/rockstor
rm -rf .venv/
rm -rf static/

userdel radmin
```

Re-build in normal mode (not DEBUG):
```
systemctl stop postgresql && systemctl start postgresql && ./build.sh && poetry run initrock && systemctl enable --now rockstor-bootstrap
```

To use the toolbar:
```
cd /opt/rockstor
poetry run debug-mode ON
poetry run django-admin collectstatic --no-input --verbosity 2 --clear
```

FroggyFlox commented:
For reference, the static files needed by the Django Debug Toolbar are as follows:
```
.venv/lib/python3.11/site-packages/debug_toolbar/static/
└── debug_toolbar
    ├── css
    │   ├── print.css
    │   └── toolbar.css
    └── js
        ├── history.js
        ├── redirect.js
        ├── timer.js
        ├── toolbar.js
        └── utils.js

4 directories, 7 files
```
... and are all copied to '/opt/rockstor/static' during `collectstatic`.

FroggyFlox commented:
I have now rebased this branch on latest `testing` and updated Django debug toolbar to the latest version.

An rpm built using this branch successfully installs but note that the toolbar is NOT immediately available. Indeed, to do so, one needs to turn ON debug mode and re-collect static files for DJango to now detect the toolbar's static files (this is because the toolbar is not listed as an `INSTALLED_APP` when `DEBUG` is `False`). This is an additional set of small manual steps but given this tool is for developer use, I still think it's ok.

I will thus soon close this PR and submit a new and proper, squashed PR.

FroggyFlox commented:
Closing as proper PR is now available: #3007.
